// bamaramabox/index.js
module.exports = function BamaramaBox(dispatch) {
	let cid;
	let name;
  let boxLoop = false;
  let itemSlot;

 dispatch.hook('sLogin', (event) => {
    cid = event.cid;
    name = event.name;
  });

  dispatch.hook('sInven', (event) => {

    // Scan inventory
    for ( var i = 0; i < event.items.length; i++) {

      // If a chef hat or wig is found, discard it.
      if (event.items[i].item === 80090 || event.items[i].item === 80089) {
          itemSlot = event.items[i].slot;

          setTimeout(function () { dispatch.toServer('cDelItem', {
            cid: cid,
            slot: itemSlot - 40,
            amount: 1
          });}, 500);
      }
    }

    // Scan inventory
    for ( var j = 0; j < event.items.length; j++) {
        if (event.items[j].item === 80086) {
          if (boxLoop) {
            setTimeout(function () {
              dispatch.toServer('cUseItem', {
                ownerId: cid,
                item: 80086,
                id: 0,
                unk1: 0,
                unk2: 0,
                unk3: 0,
                unk4: 1,
                unk5: 0,
                unk6: 0,
                unk7: 0,
                x: 0,
                y: 0,
                z: 0,
                w: 0,
                unk8: 0,
                unk9: 0,
                unk10: 0,
                unk11: 0,
              });
          }, 5000);

            return;
        }
      }
    }
  });

  dispatch.hook('cWhisper', (event) => {

   // Ingame controls for starting and stopping the script. 
   if (event.target === 'BamBoxOpener') {
      if (event.message === '<FONT>start</FONT>') {

        boxLoop = true;

  	  	dispatch.toClient('sWhisper', {
          player: cid,
          unk1: 0,
          gm: 0,
          unk2: 0,
          author: 'Bamarama Box Opener',
          recipient: name,
          message: 'Opening your bamarama boxes.' 
        });

        dispatch.toServer('cUseItem', {
            ownerId: cid,
            item: 80086,
            id: 0,
            unk1: 0,
            unk2: 0,
            unk3: 0,
            unk4: 1,
            unk5: 0,
            unk6: 0,
            unk7: 0,
            x: 0,
            y: 0,
            z: 0,
            w: 0,
            unk8: 0,
            unk9: 0,
            unk10: 0,
            unk11: 0,
          });
      } else if (event.message === '<FONT>stop</FONT>') {
        boxLoop = false;
      	dispatch.toClient('sWhisper', {
	          player: cid,
	          unk1: 0,
	          gm: 0,
	          unk2: 0,
	          author: 'Bamarama Box Opener',
	          recipient: name,
	          message: 'Box opening stopped.' 
	     });
      }
    }
  });

};
